src/interrupt/kvm/mod.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Manage virtual device's interrupts based on the Linux KVM framework.
+//!
+//! When updaing KVM IRQ routing by ioctl(KVM_SET_GSI_ROUTING), all interrupts of the virtual
+//! machine must be updated all together. The [KvmIrqRouting](struct.KvmIrqRouting.html)
+//! structure is to maintain the global interrupt routing table.
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use kvm_bindings::{kvm_irq_routing, kvm_irq_routing_entry};
+#[cfg(all(
+    feature = "legacy_irq",
+    any(target_arch = "x86", target_arch = "x86_64")
+))]
+use kvm_bindings::{
+    KVM_IRQCHIP_IOAPIC, KVM_IRQCHIP_PIC_MASTER, KVM_IRQCHIP_PIC_SLAVE, KVM_IRQ_ROUTING_IRQCHIP,
+};
+use kvm_ioctls::VmFd;
+
+use super::*;
+
+/// Structure to manage interrupt sources for a virtual machine based on the Linux KVM framework.
+///
+/// The KVM framework provides methods to inject interrupts into the target virtual machines,
+/// which uses irqfd to notity the KVM kernel module for injecting interrupts. When the interrupt
+/// source, usually a virtual device backend in userspace, writes to the irqfd file descriptor,
+/// the KVM kernel module will inject a corresponding interrupt into the target VM according to
+/// the IRQ routing configuration.
+pub struct KvmIrqManager {
+    mgr: Mutex<KvmIrqManagerObj>,
+}
+
+impl KvmIrqManager {
+    /// Create a new interrupt manager based on the Linux KVM framework.
+    ///
+    /// # Arguments
+    /// * `vmfd`: The KVM VM file descriptor, which will be used to access the KVM subsystem.
+    pub fn new(vmfd: Arc<VmFd>) -> Self {
+        let vmfd2 = vmfd.clone();
+        KvmIrqManager {
+            mgr: Mutex::new(KvmIrqManagerObj {
+                vmfd,
+                groups: HashMap::new(),
+                routes: Arc::new(KvmIrqRouting::new(vmfd2)),
+            }),
+        }
+    }
+
+    /// Prepare the interrupt manager for generating interrupts into the target VM.
+    ///
+    /// On x86 platforms, this will set up IRQ routings for legacy IRQs.
+    pub fn initialize(&self) -> Result<()> {
+        // Safe to unwrap because there's no legal way to break the mutex.
+        let mgr = self.mgr.lock().unwrap();
+        mgr.initialize()
+    }
+}
+
+impl InterruptManager for KvmIrqManager {
+    fn create_group(
+        &self,
+        ty: InterruptSourceType,
```

Comment:
I still believe that making the interrupt source type an enum (as opposed to having a separate manager for each type) is a bad idea.

Reply:
I agree, and I prefer the approach we have taken in CH here: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/vm-device/src/interrupt/mod.rs

Reply:
> I still believe that making the interrupt source type an enum (as opposed to having a separate manager for each type) is a bad idea.

I changed my mind to leave that part to each vmm, and focus on InterruptSourceGroup only.